### PR TITLE
Tools: script: uploader.py: add --identify option

### PR DESCRIPTION
**Pull request by peterbarker**
Add option to make it absolutely clear what the MCU revision is and
whether it is flawed

### CODE REVIEW

File: Tools/scripts/uploader.py [hunk 1]
```diff
@@ -786,6 +790,39 @@ def main():
                     # Go to the next port
                     continue
 
+                if args.identify:
+                    print(" Flash size: %u" % up.fw_maxsize)
+                    print(" board_type: %u" % up.board_type)
+                    print(" board_rev: %u" % up.board_rev)
+                    if up.bl_rev >= 5:
+                        chip = up.getCHIP()
+                        MCU_REV_STM32F4_REV_A = 0x1000
```
Comment (tridge) [inline Tools/scripts/uploader.py]: this won't work for F7 and H7 chips

**Comment by tridge** [comment]
why not just ask bootloader how much flash is available?

File: Tools/scripts/uploader.py (hunk 1 above)
Comment (peterbarker) [reply Tools/scripts/uploader.py]: Fixed.  Sort of.  We don't extract much information from the F7s etc.

**Comment by peterbarker** [comment]
Uploads fine.

Identifies fine on PixRacer, fmuv2 and auav x2.1 777

**Comment by peterbarker** [comment]
```
pbarker@bluebottle:~/rc/ardupilot(pr/uploader-identify-argument)$ ./Tools/scripts/uploader.py --identify
If the board does not respond within 1-2 seconds, unplug and re-plug the USB connector.
Attempting reboot on /dev/serial/by-id/usb-ArduPilot_fmuv2_210039000947323433353231-if00 with baudrate=57600...
If the board does not respond, unplug and re-plug the USB connector.
Found board 9,0 bootloader rev 5 on /dev/serial/by-id/usb-ArduPilot_Pixhawk1-BL_210039000947323433353231-if00
Bootloader Protocol: 5
OTP:
  type: PX4
  idtype: =00
  vid: 000026ac
  pid: 00000010
  coa: mqsZkB8bACPJQyDwOQFRMOqwcdW7qK3+dX/91vQ5I7h/SsaEmvpoNldFRvK9rOuJIjglnDfekrGUYOW53ze3lpKtlFkxvQciNoZKkmqmTNea/p+KJ7RTgsITJr7XBE22DGLJEfuLAEUSUSHZnTZpObeL1x8gjruswqbzl2DBwS4=
  sn: 003900213432470931323533
ChipDes:
  family: STM32F42x
  revision: A
Chip:
  10006419 STM32F42x_43x revA (flawed; 1M limit)
Info:
  flash size: 1032192
  board_type: 9
  board_rev: 0
Identification complete
pbarker@bluebottle:~/rc/ardupilot(pr/uploader-identify-argument)$ 
```

**Comment by peterbarker** [comment]
```
[547/548] apj_gen build/fmuv2/bin/arduplane.bin
[548/548] Uploading build/fmuv2/bin/arduplane.apj
Loaded firmware for 9,0, size: 985512 bytes, waiting for the bootloader...
If the board does not respond within 1-2 seconds, unplug and re-plug the USB connector.
Found board 9,0 bootloader rev 5 on /dev/serial/by-id/usb-ArduPilot_Pixhawk1-BL_210039000947323433353231-if00
Bootloader Protocol: 5
OTP:
  type: PX4
  idtype: =00
  vid: 000026ac
  pid: 00000010
  coa: mqsZkB8bACPJQyDwOQFRMOqwcdW7qK3+dX/91vQ5I7h/SsaEmvpoNldFRvK9rOuJIjglnDfekrGUYOW53ze3lpKtlFkxvQciNoZKkmqmTNea/p+KJ7RTgsITJr7XBE22DGLJEfuLAEUSUSHZnTZpObeL1x8gjruswqbzl2DBwS4=
  sn: 003900213432470931323533
ChipDes:
  family: STM32F42x
  revision: A
Chip:
  10006419 STM32F42x_43x revA (flawed; 1M limit)
Info:
  flash size: 1032192
  board_type: 9
  board_rev: 0
Identification complete

Erase  : [====================] 100.0%
Program: [====================] 100.0%
Verify : [====================] 100.0%
Rebooting.

Waf: Leaving directory `/home/pbarker/rc/ardupilot/build/fmuv2'

BUILD SUMMARY
Build directory: /home/pbarker/rc/ardupilot/build/fmuv2
Target         Text    Data  BSS     Total  
--------------------------------------------
bin/arduplane  984348  1152  195644  1181144

Build commands will be stored in build/fmuv2/compile_commands.json
'build' finished successfully (30.221s)
pbarker@bluebottle:~/rc/ardupilot(pr/uploader-identify-argument)$ 
```

File: Tools/scripts/uploader.py [hunk 2]
```diff
@@ -794,14 +851,15 @@ def main():
     parser.add_argument('--source-system', type=int, action="store", help="Source system to send reboot mavlink packets from", default=255)
     parser.add_argument('--source-component', type=int, action="store", help="Source component to send reboot mavlink packets from", default=0)
     parser.add_argument('--download', action='store_true', default=False, help='download firmware from board')
-    parser.add_argument('firmware', action="store", help="Firmware file to be uploaded")
+    parser.add_argument('--identify', action="store_true", help="Do not flash firmware; simply dump information about board")
+    parser.add_argument('firmware', nargs="?", action="store", default=None, help="Firmware file to be uploaded")
```
Comment (OXINARF) [inline Tools/scripts/uploader.py]: Given that firmware is now optional, shouldn't we add a new check that firmware is not None unless command is identify?

File: Tools/scripts/uploader.py [hunk 3]
```diff
@@ -840,6 +898,10 @@ def main():
                     # Go to the next port
                     continue
 
+                if args.identify:
+                    up.dump_board_info()
+                    sys.exit(0)
```
Comment (OXINARF) [inline Tools/scripts/uploader.py]: Does this close the port (meaning does finally part run after the exit)?

In any case, I think this should be inside the try below and just be an additional option (without the exit part):

```
if args.identify:
  up.dump_board_info()
else if args.download:
 [...]
else:
 [...]
```

**Comment by tridge** [comment]
works on H7 and F7

**Comment by tridge** [comment]
tridge@blu3:~/APM.2(pr/uploader-identify-argument)$ ./Tools/scripts/uploader.py 
Traceback (most recent call last):
  File "./Tools/scripts/uploader.py", line 936, in <module>
    main()
  File "./Tools/scripts/uploader.py", line 863, in main
    fw = firmware(args.firmware)
  File "./Tools/scripts/uploader.py", line 142, in __init__
    f = open(path, "r")
TypeError: coercing to Unicode: need string or buffer, NoneType found

File: Tools/scripts/uploader.py [hunk 4]
```diff
@@ -575,6 +573,98 @@ def identify(self):
...
+                0x413: "STM32F40x_41x",
+                0x419: "STM32F42x_43x",
+                0x421: "STM32F42x_446xx",
+            }
+            F7_IDS = {
+                0x449: "STM32F74x_75x",
+                0x451: "STM32F76x_77x",
+            }
+
+            family = mcu_id & 0xfff
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply Tools/scripts/uploader.py]: It's matching code elsewhere dealing with this - trying to make things look familiar.

File: Tools/scripts/uploader.py [hunk 5]
```diff
@@ -575,6 +573,98 @@ def identify(self):
...
+        print("Chip:")
+        if self.bl_rev > 4:
+            chip = self.__getCHIP()
+            mcu_id = chip & 0xfff
+            revs = {}
+
+            F4_IDS = {
+                0x413: "STM32F40x_41x",
+                0x419: "STM32F42x_43x",
+                0x421: "STM32F42x_446xx",
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply Tools/scripts/uploader.py]: I'm just swiping this code elsewhere - in this case here: https://github.com/ardupilot/ardupilot/blob/master/Tools/AP_Bootloader/mcu_f4.h#L17

File: Tools/scripts/uploader.py (hunk 2 above)
Comment (peterbarker) [reply Tools/scripts/uploader.py]: Yes, I've added a check, thanks!

File: Tools/scripts/uploader.py (hunk 3 above)
Comment (peterbarker) [reply Tools/scripts/uploader.py]: Done.

**Pull request merged**